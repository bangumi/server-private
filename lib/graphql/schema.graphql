type Query {
    me: User
    character(id: Int!): Character
    person(id: Int!): Person
    subject(id: Int!): Subject
}

type Avatar {
    large: String!
    medium: String!
    small: String!
}

"""basic character info as field of other types to avoid recursive query"""
type SlimCharacter {
    collects: Int!
    comment: Int!
    id: Int!
    images: Images
    infobox: [Infobox!]
    last_post: Int!
    lock: Int!
    name: String!
    nsfw: Boolean!
    redirect: Int!
    role: Int!
    summary: String!
}

type Character {
    collects: Int!
    comment: Int!
    id: Int!
    images: Images
    infobox: [Infobox!]
    last_post: Int!
    lock: Int!
    name: String!
    nsfw: Boolean!
    persons(limit: Int! = 10, offset: Int! = 0): [CharacterRelatedPerson!]
    redirect: Int!
    role: Int!
    subjects(limit: Int! = 10, offset: Int! = 0): [CharacterRelatedSubject!]
    summary: String!
}

type CharacterRelatedPerson {
    person: SlimPerson!
    subject: SlimSubject!
    summary: String!
}

type CharacterRelatedSubject {
    order: Int!
    subject: SlimSubject!
    type: Int!
}

type Episode {
    airdate: String!
    comment: Int!
    description: String!
    disc: Int!
    duration: String!
    id: Int!
    last_post: Int!
    name: String!
    name_cn: String!
    sort: Float!
    type: Int!
}

type Images {
    grid: String!
    large: String!
    medium: String!
    small: String!
}

type Infobox {
    key: String!
    values: [InfoboxValue!]
}

type InfoboxValue {
    k: String
    v: String
}

"""basic person info as field of other types to avoid recursive query"""
type SlimPerson {
    career: [String!]!
    collects: Int!
    comment: Int!
    id: Int!
    images: Images
    infobox: [Infobox!]
    last_post: Int!
    lock: Int!
    name: String!
    nsfw: Boolean!
    redirect: Int!
    summary: String!
    type: Int!
}

type Person {
    career: [String!]!
    characters(limit: Int! = 10, offset: Int! = 0): [PersonRelatedCharacter!]
    collects: Int!
    comment: Int!
    id: Int!
    images: Images
    infobox: [Infobox!]
    last_post: Int!
    lock: Int!
    name: String!
    nsfw: Boolean!
    redirect: Int!
    subjects(limit: Int! = 10, offset: Int! = 0): [PersonRelatedSubject!]
    summary: String!
    type: Int!
}

type PersonRelatedCharacter {
    character: SlimCharacter!
    subject: SlimSubject!
    summary: String!
}

type PersonRelatedSubject {
    position: Int!
    subject: SlimSubject!
}

"""a subject as field of other types to avoid recursive query"""
type SlimSubject {
    airtime: SubjectAirtime!
    eps: Int!
    id: Int!
    images: SubjectImages
    infobox: [Infobox!]
    locked: Boolean!
    name: String!
    name_cn: String!
    nsfw: Boolean!
    platform: SubjectPlatform!
    rating: SubjectRating!
    redirect: Int!
    series: Boolean!
    series_entry: Int!
    summary: String!
    type: Int!
    volumes: Int!
}

type Subject {
    airtime: SubjectAirtime!
    characters(limit: Int! = 10, offset: Int! = 0): [SubjectRelatedCharacter!]
    collection: SubjectCollection!
    episodes(
        limit: Int! = 30

        """超出范围时会返回空数据。可以使用负数，来获取最后几个章节。如 `-1` 会返回最后一个章节， `-2` 会返回最后两个章节"""
        offset: Int! = 0
        type: Int
    ): [Episode!]
    eps: Int!
    id: Int!
    images: SubjectImages
    infobox: [Infobox!]
    locked: Boolean!
    name: String!
    name_cn: String!
    nsfw: Boolean!
    persons(limit: Int! = 10, offset: Int! = 0): [SubjectRelatedPerson!]
    platform: SubjectPlatform!
    rating: SubjectRating!
    redirect: Int!
    relations(excludeTypes: [Int!], includeTypes: [Int!], limit: Int! = 30, offset: Int! = 0): [SubjectRelation!]
    series: Boolean!
    series_entry: Int!
    summary: String!
    tags(limit: Int = 0): [SubjectTag!]!
    topics(limit: Int! = 10, offset: Int! = 0): [SubjectTopic!]
    type: Int!
    volumes: Int!
}

type SubjectAirtime {
    date: String!
    month: Int!
    weekday: Int!
    year: Int!
}

type SubjectCollection {
    collect: Int!
    doing: Int!
    dropped: Int!
    on_hold: Int!
    wish: Int!
}

type SubjectImages {
    common: String!
    grid: String!
    large: String!
    medium: String!
    small: String!
}

type SubjectPlatform {
    alias: String
    id: Int!
    type: String!
    type_cn: String
}

type SubjectRating {
    count: [Int!]
    rank: Int!
    score: Float!
    total: Int!
}

type SubjectRelatedCharacter {
    character: SlimCharacter!
    order: Int!
    type: Int!
}

type SubjectRelatedPerson {
    person: SlimPerson!
    position: Int!
}

type SubjectRelation {
    order: Int!
    relation: Int!
    subject: SlimSubject!
}

type SubjectTag {
    count: Int!
    name: String!
}

type SubjectTopic {
    created_at: Int!
    creator: User!
    display: Int!
    id: Int!
    replies: Int!
    state: Int!
    title: String!
    updated_at: Int!
}

type User {
    avatar: Avatar!
    id: Int!
    nickname: String!
    username: String!
}
