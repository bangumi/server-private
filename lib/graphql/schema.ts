import * as fsp from 'node:fs/promises';
import * as path from 'node:path';

import { projectRoot } from '@app/lib/config.ts';
import type { Context } from '@app/lib/graphql/context.ts';
import { convertCharacter } from '@app/lib/graphql/resolvers/character.ts';
import { convertPerson } from '@app/lib/graphql/resolvers/person.ts';
import {
  convertSubject,
  convertTopic,
  subjectResolver,
} from '@app/lib/graphql/resolvers/subject.ts';
import * as entity from '@app/lib/orm/entity/index.ts';
import {
  CastRepo,
  CharacterRepo,
  CharacterSubjectsRepo,
  fetchUser,
  PersonRepo,
  PersonSubjectsRepo,
  SubjectRelationRepo,
  SubjectTopicRepo,
} from '@app/lib/orm/index.ts';
import { avatar } from '@app/lib/response.ts';
import { ListTopicDisplays } from '@app/lib/topic/index.ts';

import type * as types from './__generated__/resolvers.ts';

export function convertUser(user: entity.User) {
  return {
    id: user.id,
    username: user.username,
    nickname: user.nickname,
    avatar: avatar(user.avatar),
  };
}

export const resolvers = {
  Query: {
    async me(_parent, _args, { auth }: Context) {
      if (!auth.userID) {
        return null;
      }

      const user = await fetchUser(auth.userID);
      if (!user) {
        return null;
      }

      return {
        id: user.id,
        avatar: avatar(user.img),
        nickname: user.nickname,
        username: user.username,
      };
    },

    subject: subjectResolver,

    async person(_parent, { id }: { id: number }, { auth: { allowNsfw } }: Context) {
      let query = PersonRepo.createQueryBuilder('c').where('c.id = :id', { id });
      if (!allowNsfw) {
        query = query.andWhere('c.nsfw = :allowNsfw', { allowNsfw });
      }
      query = query.andWhere('c.ban = 0');
      const person = await query.getOne();
      if (!person) {
        return null;
      }
      return convertPerson(person);
    },

    async character(_parent, { id }: { id: number }, { auth: { allowNsfw } }: Context) {
      let query = CharacterRepo.createQueryBuilder('c').where('c.id = :id', { id });
      if (!allowNsfw) {
        query = query.andWhere('c.nsfw = :allowNsfw', { allowNsfw });
      }
      query = query.andWhere('c.ban = 0');
      const character = await query.getOne();
      if (!character) {
        return null;
      }
      return convertCharacter(character);
    },
  },

  Subject: {
    tags(parent: types.Subject, { limit }): types.SubjectTag[] {
      if (limit && limit > 0) {
        return parent.tags.slice(0, limit);
      }

      return parent.tags;
    },

    async episodes(
      parent: { id: number },
      { limit, offset, type },
      { repo }: Context,
    ): Promise<types.Episode[]> {
      if (!parent.id) {
        return [];
      }

      if (offset < 0) {
        const count = await repo.Episode.createQueryBuilder('t')
          .where('t.subjectID = :s', { s: parent.id })
          .getCount();

        offset = count + offset;
      }

      let s = repo.Episode.createQueryBuilder('t')
        .select()
        .where('t.subjectID = :s', { s: parent.id });
      if (type) {
        s = s.andWhere('t.type = :t', { t: type });
      }
      const episodes = await s.orderBy('t.sort', 'ASC').limit(limit).offset(offset).getMany();

      return episodes.map((e: entity.Episode) => {
        return {
          id: e.id,
          type: e.type,
          name: e.name,
          name_cn: e.nameCN,
          description: e.summary,
          airdate: e.airDate,
          comment: e.epComment,
          last_post: e.epLastPost,
          disc: e.epDisc,
          duration: e.duration,
          sort: e.sort,
        };
      });
    },

    async topics(
      parent: { id: number },
      { limit, offset }: { limit: number; offset: number },
      { auth: u }: Context,
    ) {
      let query = SubjectTopicRepo.createQueryBuilder('t')
        .innerJoinAndMapOne('t.creator', entity.User, 'u', 'u.id = t.creatorID')
        .where('t.parentID = :id', { id: parent.id });
      const displays = ListTopicDisplays(u);
      if (displays.length > 0) {
        query = query.andWhere('t.display IN (:...displays)', { displays });
      }
      const topics = await query.orderBy('t.createdAt', 'DESC').skip(offset).take(limit).getMany();
      return topics.map((t) => {
        return convertTopic(t);
      });
    },

    async persons(
      parent: { id: number },
      { limit, offset }: { limit: number; offset: number },
      { auth: { allowNsfw } }: Context,
    ) {
      let query = PersonSubjectsRepo.createQueryBuilder('r')
        .innerJoinAndMapOne('r.person', entity.Person, 'p', 'p.id = r.personID')
        .where('r.subjectID = :id', { id: parent.id });
      if (!allowNsfw) {
        query = query.andWhere('p.nsfw = :allowNsfw', { allowNsfw });
      }
      const relations = await query.orderBy('r.position', 'ASC').skip(offset).take(limit).getMany();
      return relations.map((r) => {
        return {
          person: convertPerson(r.person),
          position: r.position,
        };
      });
    },

    async characters(
      parent: { id: number },
      { limit, offset }: { limit: number; offset: number },
      { auth: { allowNsfw } }: Context,
    ) {
      let query = CharacterSubjectsRepo.createQueryBuilder('r')
        .innerJoinAndMapOne('r.character', entity.Character, 'c', 'c.id = r.characterID')
        .where('r.subjectID = :id', { id: parent.id });
      if (!allowNsfw) {
        query = query.andWhere('c.nsfw = :allowNsfw', { allowNsfw });
      }
      const relations = await query
        .orderBy('r.type', 'ASC')
        .orderBy('r.order', 'ASC')
        .skip(offset)
        .take(limit)
        .getMany();
      return relations.map((r) => {
        return {
          character: convertCharacter(r.character),
          type: r.type,
          order: r.order,
        };
      });
    },

    async relations(
      parent: { id: number },
      { limit, offset, includeTypes, excludeTypes },
      { auth: { allowNsfw } }: Context,
    ): Promise<types.SubjectRelation[]> {
      let query = SubjectRelationRepo.createQueryBuilder('r')
        .innerJoinAndMapOne('r.relatedSubject', entity.Subject, 's', 's.id = r.relatedSubjectID')
        .innerJoinAndMapOne('s.fields', entity.SubjectFields, 'f', 'f.subjectID = s.id')
        .where('r.subjectID = :id', { id: parent.id });
      if (includeTypes && includeTypes.length > 0) {
        query = query.andWhere('r.relationType IN (:...includeTypes)', { includeTypes });
      }
      if (excludeTypes && excludeTypes.length > 0) {
        query = query.andWhere('r.relationType NOT IN (:...excludeTypes)', { excludeTypes });
      }
      if (!allowNsfw) {
        query = query.andWhere('s.subjectNsfw = :allowNsfw', { allowNsfw });
      }
      const relations = await query
        .orderBy('r.relationType', 'ASC')
        .orderBy('r.order', 'ASC')
        .orderBy('r.relatedSubjectID', 'ASC')
        .skip(offset)
        .take(limit)
        .getMany();
      return relations.map((r: entity.SubjectRelation) => {
        return {
          subject: convertSubject(r.relatedSubject),
          relation: r.relationType,
          order: r.order,
        };
      });
    },
  },
  Character: {
    async subjects(
      parent: { id: number },
      { limit, offset }: { limit: number; offset: number },
      { auth: { allowNsfw } }: Context,
    ) {
      let query = CharacterSubjectsRepo.createQueryBuilder('r')
        .innerJoinAndMapOne('r.subject', entity.Subject, 's', 's.id = r.subjectID')
        .innerJoinAndMapOne('s.fields', entity.SubjectFields, 'f', 'f.subjectID = s.id')
        .where('r.characterID = :id', { id: parent.id });
      if (!allowNsfw) {
        query = query.andWhere('s.subjectNsfw = :allowNsfw', { allowNsfw });
      }
      const relations = await query
        .orderBy('r.type', 'ASC')
        .orderBy('r.order', 'ASC')
        .skip(offset)
        .take(limit)
        .getMany();
      return relations.map((r) => ({
        subject: convertSubject(r.subject),
        type: r.type,
        order: r.order,
      }));
    },
    async persons(
      parent: { id: number },
      { limit, offset }: { limit: number; offset: number },
      { auth: { allowNsfw } }: Context,
    ) {
      const query = CastRepo.createQueryBuilder('r')
        .innerJoinAndMapOne('r.person', entity.Person, 'p', 'p.id = r.personID')
        .innerJoinAndMapOne('r.subject', entity.Subject, 's', 's.id = r.subjectID')
        .innerJoinAndMapOne('s.fields', entity.SubjectFields, 'f', 'f.subjectID = s.id')
        .where('r.characterID = :id', { id: parent.id });
      if (!allowNsfw) {
        query.andWhere('s.subjectNsfw = :allowNsfw', { allowNsfw });
      }
      const relations = await query.skip(offset).take(limit).getMany();
      return relations.map((r) => ({
        person: convertPerson(r.person),
        subject: convertSubject(r.subject),
        summary: r.summary,
      }));
    },
  },
  Person: {
    async characters(
      parent: { id: number },
      { limit, offset }: { limit: number; offset: number },
      { auth: { allowNsfw } }: Context,
    ) {
      const query = CastRepo.createQueryBuilder('r')
        .innerJoinAndMapOne('r.character', entity.Character, 'c', 'c.id = r.characterID')
        .innerJoinAndMapOne('r.subject', entity.Subject, 's', 's.id = r.subjectID')
        .innerJoinAndMapOne('s.fields', entity.SubjectFields, 'f', 'f.subjectID = s.id')
        .where('r.personID = :id', { id: parent.id });
      if (!allowNsfw) {
        query.andWhere('s.subjectNsfw = :allowNsfw', { allowNsfw });
      }
      const relations = await query.skip(offset).take(limit).getMany();
      return relations.map((r) => ({
        character: convertCharacter(r.character),
        subject: convertSubject(r.subject),
        summary: r.summary,
      }));
    },

    async subjects(
      parent: { id: number },
      { limit, offset }: { limit: number; offset: number },
      { auth: { allowNsfw } }: Context,
    ) {
      let query = PersonSubjectsRepo.createQueryBuilder('r')
        .innerJoinAndMapOne('r.subject', entity.Subject, 's', 's.id = r.subjectID')
        .innerJoinAndMapOne('s.fields', entity.SubjectFields, 'f', 'f.subjectID = s.id')
        .where('r.personID = :id', { id: parent.id });
      if (!allowNsfw) {
        query = query.andWhere('s.subjectNsfw = :allowNsfw', { allowNsfw });
      }
      const relations = await query.orderBy('r.position', 'ASC').skip(offset).take(limit).getMany();
      return relations.map((r) => ({
        subject: convertSubject(r.subject),
        position: r.position,
      }));
    },
  },
} satisfies types.Resolvers;

export const schema = await fsp.readFile(
  path.join(projectRoot, 'lib', 'graphql', 'schema.graphql'),
  'utf8',
);
