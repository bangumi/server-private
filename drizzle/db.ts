import type { ExtractTablesWithRelations } from 'drizzle-orm';
import type { MySqlTransaction } from 'drizzle-orm/mysql-core';
import {
  drizzle,
  type MySql2PreparedQueryHKT,
  type MySql2QueryResultHKT,
} from 'drizzle-orm/mysql2';
import * as mysql from 'mysql2/promise';

import config, { developing, testing } from '@app/lib/config.ts';
import { logger } from '@app/lib/logger.ts';

import * as schema from './schema.ts';

const poolConnection = mysql.createPool({
  host: config.mysql.host,
  user: config.mysql.user,
  database: config.mysql.db,
  password: config.mysql.password,
  port: config.mysql.port,
});

export const db = drizzle(poolConnection, {
  schema,
  mode: 'default',
  logger:
    testing || developing
      ? {
          logQuery(query: string, params: unknown[]) {
            logger.trace('query', { query, params });
          },
        }
      : undefined,
});

export * as op from 'drizzle-orm';

export type Txn = MySqlTransaction<
  MySql2QueryResultHKT,
  MySql2PreparedQueryHKT,
  typeof schema,
  ExtractTablesWithRelations<Record<string, unknown>>
>;
