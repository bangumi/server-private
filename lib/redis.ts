import * as url from 'node:url';

import type { RedisOptions } from 'ioredis';
import IORedis from 'ioredis';

import config from './config';

const u = url.parse(config.redis.uri);

const [username, password] = (u.auth ?? '').split(':', 2);

export const redisOption = {
  host: u.hostname ?? '127.0.0.1',
  port: u.port ? Number.parseInt(u.port) : 3306,
  db: u.pathname ? Number.parseInt(u.pathname.slice(1)) : 0,
  username,
  password,
  lazyConnect: true,
} satisfies RedisOptions;

export default new IORedis(redisOption);
export const Subscriber = new IORedis(redisOption);
