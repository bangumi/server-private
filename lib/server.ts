import fastifyRequestContextPlugin from '@fastify/request-context';
import type { Static } from '@sinclair/typebox';
import type Ajv from 'ajv';
import addFormats from 'ajv-formats';
import { DrizzleError } from 'drizzle-orm';
import type { FastifyInstance, FastifyRequest, FastifyServerOptions } from 'fastify';
import { fastify } from 'fastify';
import type { FastifySchemaValidationError } from 'fastify/types/schema.d.ts';
import metricsPlugin from 'fastify-metrics';
import mercurius from 'mercurius';
import { TypeORMError } from 'typeorm';

import { fastifyAltairPlugin } from '@app/lib/graphql/ui.ts';
import * as routes from '@app/routes/index.ts';

import { emptyAuth } from './auth/index.ts';
import * as auth from './auth/index.ts';
import config, { testing, VERSION } from './config.ts';
import type { Context } from './graphql/context.ts';
import { resolvers, schema } from './graphql/schema.ts';
import { repo } from './orm/index.ts';
import type * as res from './types/res.ts';

export function defaultSchemaErrorFormatter(
  errors: FastifySchemaValidationError[],
  dataVar: string,
): Error & Static<typeof res.Error> {
  let text = '';
  const separator = ', ';

  for (const e of errors) {
    text += `${dataVar}${e.instancePath} ${e.message ?? ''}${separator}`;
  }
  return new ValidationError(text.slice(0, -separator.length));
}

class ValidationError extends Error {
  code = 'REQUEST_VALIDATION_ERROR';
  error = 'Bad Request';
  statusCode = 400;

  constructor(msg: string) {
    super(msg);
    this.message = msg;
  }
}

declare module '@fastify/request-context' {
  interface RequestContextData {
    req: unknown;
    user?: number;
  }
}

export async function createServer(
  opts: Omit<
    FastifyServerOptions,
    'onProtoPoisoning' | 'ajv' | 'schemaErrorFormatter' | 'requestIdHeader'
  > = {},
): Promise<FastifyInstance> {
  const ajv: FastifyServerOptions['ajv'] = {
    plugins: [
      function (ajv: Ajv) {
        ajv.addKeyword({ keyword: 'x-examples' });
        ajv.addKeyword({ keyword: 'x-ms-enum' });
        ajv.addKeyword({ keyword: 'x-enum-varnames' });
      },
      addFormats,
    ],
  };

  const server = fastify({
    ...opts,
    onProtoPoisoning: 'error',
    ajv,
    bodyLimit: 8 * 1024 * 1024,
    requestIdHeader: config.server.requestIDHeader,
    schemaErrorFormatter: defaultSchemaErrorFormatter,
  });

  server.setErrorHandler(function (error, request, reply) {
    // hide TypeORM message
    if (error instanceof TypeORMError || error instanceof DrizzleError) {
      this.log.error(error);
      void reply.status(500).send({
        error: 'Internal Server Error',
        message: 'internal database error, please contact admin',
        statusCode: 500,
      });
    } else {
      if (typeof error.statusCode !== 'number' || error.statusCode === 500) {
        this.log.error(error);
        void reply.status(500).send({
          error: 'Internal Server Error',
          message: 'internal error, please contact admin',
          statusCode: 500,
        });
      } else {
        void reply.send(error);
      }
    }
  });

  server.addHook('onRequest', (req, res, done) => {
    void res.header('x-server-version', VERSION);
    done();
  });

  const clientIpHeader = config.server.clientIpHeader;

  server.decorateRequest('ip', {
    getter: function (this: FastifyRequest): string {
      const hRealIp = this.headers[clientIpHeader] as string | undefined;
      if (hRealIp) {
        return hRealIp;
      }

      return this.socket.remoteAddress ?? '127.0.0.1';
    },
  });

  if (!testing) {
    await server.register(metricsPlugin, {
      routeMetrics: {
        groupStatusCodes: false,
        overrides: {
          histogram: {
            buckets: [0.05, 0.1, 0.2, 0.3, 0.5, 0.75, 1, 2, 3],
          },
        },
      },
    });
  }

  await server.register(fastifyRequestContextPlugin, {
    defaultStoreValues: (req) => ({
      req: {
        id: req.id,
        url: req.url,
      },
    }),
  });

  await server.register(mercurius, {
    schema,
    resolvers,
    path: '/v0/graphql',
    jit: 200,
    queryDepth: 5,
    allowBatchedQueries: true,
    context: async (request: FastifyRequest): Promise<Context> => {
      const a = await auth.byHeader(request.headers.authorization);
      if (a) {
        return { repo, auth: a };
      }

      return { repo, auth: emptyAuth() };
    },
  });

  await server.register(fastifyAltairPlugin, {
    path: '/v0/altair/',
    baseURL: '/v0/altair/',
    endpointURL: '/v0/graphql',
    initialSettings: {
      theme: 'dark',
      'plugin.list': ['altair-graphql-plugin-graphql-explorer'],
    },
  });

  await server.register(routes.setup);

  return server;
}
