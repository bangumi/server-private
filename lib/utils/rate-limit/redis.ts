import * as fs from 'node:fs';
import * as path from 'node:path';

import type { Redis } from 'ioredis';

import { projectRoot, redisPrefix } from '@app/lib/config.ts';
import type { LimitAction, Limiter, Result } from '@app/lib/utils/rate-limit/index.ts';

const luaScript = fs.readFileSync(
  path.join(projectRoot, 'lib/utils/rate-limit/redis-script.lua'),
  'utf8',
);

export class RedisLimiter implements Limiter {
  private readonly redisClient: Redis;

  constructor({ redisClient }: { redisClient: Redis }) {
    this.redisClient = redisClient;

    this.redisClient.defineCommand('getRateLimit', {
      numberOfKeys: 1,
      lua: luaScript,
    });
  }

  async get(key: string, timeWindow: number, limit: number): Promise<Result> {
    const [remain, reset] = await this.redisClient.getRateLimit(key, limit, timeWindow);
    return {
      limited: remain < 0,
      remain,
      limit,
      reset,
    };
  }

  async userAction(
    userID: number,
    action: LimitAction,
    timeWindow: number,
    limit: number,
  ): Promise<Result> {
    const limitKey = `${redisPrefix}:rate-limit:${action}:${userID}`;

    return this.get(limitKey, timeWindow, limit);
  }

  async reset(key: string): Promise<void> {
    await this.redisClient.del(key);
  }
}

// work with defineCommand
declare module 'ioredis' {
  interface RedisCommander {
    getRateLimit(key: string, limit: number, duration: number): Promise<[number, number]>;
  }
}
