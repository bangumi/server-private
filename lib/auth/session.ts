import type { CookieSerializeOptions } from '@fastify/cookie';
import { DateTime } from 'luxon';

import { db, op, schema } from '@app/drizzle';
import { TypedCache } from '@app/lib/cache.ts';
import { randomBytes } from '@app/lib/utils/index.ts';

import type { IAuth } from './index.ts';
import * as auth from './index.ts';

export const CookieKey = 'chiiNextSessionID';
export const LegacyCookieKey = 'chii_auth';
export const cookiesPluginOption: Readonly<CookieSerializeOptions> = {
  sameSite: 'lax',
  secure: 'auto',
  path: '/',
  httpOnly: true,
};

export async function create(user: { id: number; regTime: number }): Promise<string> {
  const now = DateTime.now().toUnixInteger();
  const bytes = await randomBytes(30);
  const token = bytes.toString('base64url');

  const value = {
    reg_time: DateTime.now().toISOTime(),
    user_id: user.id,
    created_at: now,
    expired_at: now + 60 * 60 * 24 * 7,
  };

  await db.insert(schema.chiiOsWebSessions).values({
    value: Buffer.from(JSON.stringify(value)),
    userID: user.id,
    createdAt: value.created_at,
    expiredAt: value.expired_at,
    key: token,
  });

  return token;
}

interface ICachedSession {
  userID: number;
}

const sessionCache = TypedCache<string, ICachedSession>((sessionID) => `auth:session:${sessionID}`);

/** @param sessionID - Store in user cookies */
export async function get(sessionID: string): Promise<IAuth | null> {
  const cached = await sessionCache.get(sessionID);
  if (cached) {
    return await auth.byUserID(cached.userID);
  }

  const session = await db.query.chiiOsWebSessions.findFirst({
    where: op.and(
      op.sql`\`key\` = ${sessionID} collate utf8mb4_bin`,
      op.gt(schema.chiiOsWebSessions.expiredAt, DateTime.now().toUnixInteger()),
    ),
  });
  if (!session) {
    return null;
  }

  await sessionCache.set(sessionID, { userID: session.userID });

  return await auth.byUserID(session.userID);
}

export async function revoke(sessionID: string) {
  await db
    .update(schema.chiiOsWebSessions)
    .set({ expiredAt: DateTime.now().toUnixInteger() })
    .where(op.eq(schema.chiiOsWebSessions.key, sessionID));

  await sessionCache.del(sessionID);
}
