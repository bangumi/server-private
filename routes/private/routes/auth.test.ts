import { afterEach, beforeEach, describe, expect, test } from 'vitest';

import { db } from '@app/drizzle/db.ts';
import { chiiOsWebSessions } from '@app/drizzle/schema.ts';
import * as session from '@app/lib/auth/session.ts';
import redis from '@app/lib/redis.ts';
import { createServer } from '@app/lib/server.ts';

describe('login', () => {
  beforeEach(async () => {
    await redis.flushdb('SYNC');
    await db.delete(chiiOsWebSessions);
  });

  afterEach(async () => {
    await redis.flushdb('SYNC');
    await db.delete(chiiOsWebSessions);
  });

  test('should failed on too many requests', async () => {
    const app = await createServer();

    const opt = {
      method: 'post',
      url: '/p1/login',
      payload: { email: 'ee', password: 'eepp', 'cf-turnstile-response': 'fake-response' },
    } as const;

    const login = () => app.inject(opt);

    const all = await Promise.all(Array.from({ length: 20 }).map(() => login()));

    const res = await login();

    expect(all.filter((x) => x.statusCode === 429)).not.toHaveLength(0);

    expect(res.statusCode).toBe(429);
    expect(res.json()).toMatchSnapshot();
  });

  test('should login', async () => {
    const app = await createServer();

    const res = await app.inject({
      method: 'post',
      url: '/p1/login',
      payload: {
        email: 'treeholechan@gmail.com',
        password: 'lovemeplease',
        'cf-turnstile-response': 'fake-response',
      },
    });

    expect(res.statusCode).toBe(200);
    expect(res.cookies.filter((x) => x.name === session.CookieKey)).toHaveLength(1);
    expect(res.json()).toMatchSnapshot();
  });
});
