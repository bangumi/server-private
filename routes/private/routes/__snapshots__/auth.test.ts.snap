// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`login > should failed on too many requests 1`] = `
Object {
  "code": "TOO_MANY_REQUESTS",
  "error": "Too Many Requests",
  "message": "too many failed login attempts",
  "statusCode": 429,
}
`;

exports[`login > should login 1`] = `
Object {
  "avatar": Object {
    "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
    "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
    "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
  },
  "group": 10,
  "id": 382951,
  "joinedAt": 1512262276,
  "nickname": "树洞酱",
  "sign": "treeholechan@gmail.com 密码:lovemeplease",
  "username": "382951",
}
`;
